### PR TITLE
ISPN-8736 REST Authz

### PR DESCRIPTION
**Pull request by tristantarrant**
ISPN-8546 DIGEST authentication
ISPN-10374 CLIENT_CERT authentication
ISPN-8736 REST authorization

https://issues.jboss.org/browse/ISPN-8546
https://issues.jboss.org/browse/ISPN-10374
https://issues.jboss.org/browse/ISPN-8736

This will conflict with @gustavonalle 's non-blocking REST PR https://github.com/infinispan/infinispan/pull/7099

### CODE REVIEW

**Comment by gustavonalle** [comment]
Can you please rebase @tristantarrant ?

**Comment by tristantarrant** [comment]
@gustavonalle now fully non-blocking

File: server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java [hunk 1]
```diff
@@ -27,24 +30,23 @@ public BasicAuthenticator(SecurityDomain domain, String realm) {
...
+      if (auth != null && auth.length() > 5) {
          String type = auth.substring(0, 5);
          type = type.toLowerCase();
          if ("basic".equals(type)) {
             String cookie = auth.substring(6);
             cookie = new String(Base64.getDecoder().decode(cookie.getBytes()));
             String[] split = cookie.split(":");
             try {
-               request.setPrincipal(domain.authenticate(split[0], split[1]));
+               request.setSubject(domain.authenticate(split[0], split[1]));
```
Comment (gustavonalle) [inline server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java]: It seems the ```challenge()``` method will still block on ```domain.authenticate()```?
This guy takes a large chunk of the REST request processing time I'm afraid.

Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java]: Right, I'll wrap that

File: server/rest/src/main/java/org/infinispan/rest/authentication/Authenticator.java [hunk 2]
```diff
@@ -12,13 +16,29 @@
  */
 public interface Authenticator {
 
+   CompletionStage<RestResponse> VOID_RESPONSE = CompletableFuture.completedFuture(new RestResponse() {
+      @Override
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/authentication/Authenticator.java]: Done

File: server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java (hunk 1 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java]: It's now wrapped in a CompletableFuture.supplyAsync

**Comment by gustavonalle** [comment]
Integrated, thanks @tristantarrant !

**Pull request merged**